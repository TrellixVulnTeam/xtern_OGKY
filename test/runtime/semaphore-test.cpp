// RUN: %srcroot/test/runtime/run-scheduler-test.py %s -gxx "%gxx" -llvmgcc "%llvmgcc" -projbindir "%projbindir" -ternruntime "%ternruntime"  -ternbcruntime "%ternbcruntime"

#include <sys/time.h>
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <semaphore.h>
#include <assert.h>
#include <unistd.h>
#include <errno.h>

#define N (2)

sem_t sem, another_sem;
pthread_mutex_t m = PTHREAD_MUTEX_INITIALIZER;

void* sem_wait_func(void *arg) {
  char buf[64];
  int tid = (intptr_t)arg;

  for(unsigned i=0;i<10;++i)
    sched_yield();

  int ret;
  struct timespec   ts;
  struct timeval    tp;

  // test 3 different ways of sem_wait
  switch(tid % 3) {
  case 0:
    sem_wait(&sem);
    break;
  case 1:
    do{
      ret = sem_trywait(&sem);
    } while(ret!=0);
    break;
  case 2:
    do {
      gettimeofday(&tp, NULL);
      ts.tv_sec  = tp.tv_sec;
      ts.tv_nsec = tp.tv_usec * 1000 + 100;
      ret = sem_timedwait(&sem, &ts);
    } while(ret!=0);
    break;
  }

  printf("%03d WAKES UP\n", tid);
  printf("%03d POSTS\n", tid);

  sem_post(&another_sem);
}

void* sem_post_func(void *arg) {
  char buf[64];
  int tid = (intptr_t)arg;

  for(unsigned i=0;i<100;++i)
    sched_yield();

  printf("%03d POSTS\n", tid);
  sem_post(&sem);
  sem_wait(&another_sem);
  printf("%03d WAKES UP\n", tid);
}

int main(int argc, char *argv[], char *env[]) {
  int ret;
  pthread_t th[N];

  sem_init(&sem, 0, 0);
  sem_init(&another_sem, 0, 0);

  for(unsigned i=0; i<N/2; ++i) {
    ret  = pthread_create(&th[i], NULL, sem_wait_func, (void*)i);
    assert(!ret && "pthread_create() failed!");
  }

  for(unsigned i=(N+1)/2; i<N; ++i) {
    ret  = pthread_create(&th[i], NULL, sem_post_func, (void*)i);
    assert(!ret && "pthread_create() failed!");
  }

  for(unsigned i=0; i<N; ++i)
    pthread_join(th[i], NULL);

  sem_destroy(&sem);
}

// OUT indicates expected output checked by FileCheck; auto-generated by appending -gen to the RUN command above.
// OUT:      001 POSTS
// OUT-NEXT: 000 WAKES UP
// OUT-NEXT: 000 POSTS
// OUT-NEXT: 001 WAKES UP
// RR indicates expected RR schedule checked by FileCheck; auto-generated by appending -gen to the RUN command above.
// RR:      pthread_create 1 0 0
// RR-NEXT: pthread_create 3 0 0
// RR-NEXT: sem_post 5 2
// RR-NEXT: sem_wait 6 1
// RR-NEXT: sem_post 7 1
// RR-NEXT: sem_wait 8 2
// RR-NEXT: pthread_join 11 0
// RR-NEXT: pthread_join 12 0
