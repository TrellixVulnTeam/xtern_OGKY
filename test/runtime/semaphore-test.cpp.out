000 WAITS
001 WAITS
002 WAITS
003 WAITS
004 WAITS
005 WAITS
006 WAITS
007 WAITS
008 WAITS
009 WAITS
010 WAITS
011 WAITS
012 WAITS
013 WAITS
014 WAITS
015 WAITS
016 WAITS
017 WAITS
018 WAITS
019 WAITS
020 WAITS
021 WAITS
022 WAITS
023 WAITS
024 WAITS
025 WAITS
026 WAITS
027 WAITS
028 WAITS
029 WAITS
030 WAITS
031 WAITS
032 WAITS
033 WAITS
034 WAITS
035 WAITS
036 WAITS
037 WAITS
038 WAITS
039 WAITS
040 WAITS
041 WAITS
042 WAITS
043 WAITS
044 WAITS
045 WAITS
046 WAITS
047 WAITS
048 WAITS
049 WAITS
050 WAITS
051 WAITS
052 WAITS
053 WAITS
054 WAITS
055 WAITS
056 WAITS
057 WAITS
058 WAITS
059 WAITS
060 WAITS
061 WAITS
062 WAITS
063 WAITS
064 WAITS
065 WAITS
066 WAITS
067 WAITS
068 WAITS
069 WAITS
070 WAITS
071 WAITS
072 WAITS
073 WAITS
074 WAITS
075 WAITS
076 WAITS
077 WAITS
078 WAITS
079 WAITS
080 WAITS
081 WAITS
082 WAITS
083 WAITS
084 WAITS
085 WAITS
086 WAITS
087 WAITS
088 WAITS
089 WAITS
090 WAITS
091 WAITS
092 WAITS
093 WAITS
094 WAITS
095 WAITS
096 WAITS
097 WAITS
098 WAITS
099 WAITS
100 WAITS
101 WAITS
102 WAITS
103 WAITS
104 WAITS
105 WAITS
106 WAITS
107 WAITS
108 WAITS
109 WAITS
110 WAITS
111 WAITS
112 WAITS
113 WAITS
114 WAITS
115 WAITS
116 WAITS
117 WAITS
118 WAITS
119 WAITS
120 WAITS
121 WAITS
122 WAITS
123 WAITS
124 WAITS
125 WAITS
126 WAITS
127 WAITS
128 WAITS
129 WAITS
130 WAITS
131 WAITS
132 WAITS
133 WAITS
134 WAITS
135 WAITS
136 WAITS
137 WAITS
138 WAITS
139 WAITS
140 WAITS
141 WAITS
142 WAITS
143 WAITS
144 WAITS
145 WAITS
146 WAITS
147 WAITS
148 WAITS
149 WAITS
150 WAITS
151 WAITS
152 WAITS
153 WAITS
154 WAITS
155 WAITS
156 WAITS
157 WAITS
158 WAITS
159 WAITS
160 WAITS
161 WAITS
162 WAITS
163 WAITS
164 WAITS
165 WAITS
166 WAITS
167 WAITS
168 WAITS
169 WAITS
170 WAITS
171 WAITS
172 WAITS
173 WAITS
174 WAITS
175 WAITS
176 WAITS
177 WAITS
178 WAITS
179 WAITS
180 WAITS
181 WAITS
182 WAITS
183 WAITS
184 WAITS
185 WAITS
186 WAITS
187 WAITS
188 WAITS
189 WAITS
190 WAITS
191 WAITS
192 WAITS
193 WAITS
194 WAITS
195 WAITS
196 WAITS
197 WAITS
198 WAITS
199 WAITS
200 WAITS
201 WAITS
202 WAITS
203 WAITS
204 WAITS
205 WAITS
206 WAITS
207 WAITS
208 WAITS
209 WAITS
210 WAITS
211 WAITS
212 WAITS
213 WAITS
214 WAITS
215 WAITS
216 WAITS
217 WAITS
218 WAITS
219 WAITS
220 WAITS
221 WAITS
222 WAITS
223 WAITS
224 WAITS
225 WAITS
226 WAITS
227 WAITS
228 WAITS
229 WAITS
230 WAITS
231 WAITS
232 WAITS
233 WAITS
234 WAITS
235 WAITS
236 WAITS
237 WAITS
238 WAITS
239 WAITS
240 WAITS
241 WAITS
242 WAITS
243 WAITS
244 WAITS
245 WAITS
246 WAITS
247 WAITS
248 WAITS
249 WAITS
250 WAITS
251 WAITS
252 WAITS
253 WAITS
254 WAITS
255 WAITS
256 WAITS
257 WAITS
258 WAITS
259 WAITS
260 WAITS
261 WAITS
262 WAITS
263 WAITS
264 WAITS
265 WAITS
266 WAITS
267 WAITS
268 WAITS
269 WAITS
270 WAITS
271 WAITS
272 WAITS
273 WAITS
274 WAITS
275 WAITS
276 WAITS
277 WAITS
278 WAITS
279 WAITS
280 WAITS
281 WAITS
282 WAITS
283 WAITS
284 WAITS
285 WAITS
286 WAITS
287 WAITS
288 WAITS
289 WAITS
290 WAITS
291 WAITS
292 WAITS
293 WAITS
294 WAITS
295 WAITS
296 WAITS
297 WAITS
298 WAITS
299 WAITS
300 WAITS
301 WAITS
302 WAITS
303 WAITS
304 WAITS
305 WAITS
306 WAITS
307 WAITS
308 WAITS
309 WAITS
310 WAITS
311 WAITS
312 WAITS
313 WAITS
314 WAITS
315 WAITS
316 WAITS
317 WAITS
318 WAITS
319 WAITS
320 WAITS
321 WAITS
322 WAITS
323 WAITS
324 WAITS
325 WAITS
326 WAITS
327 WAITS
328 WAITS
329 WAITS
330 WAITS
331 WAITS
332 WAITS
333 WAITS
334 WAITS
335 WAITS
336 WAITS
337 WAITS
338 WAITS
339 WAITS
340 WAITS
341 WAITS
342 WAITS
343 WAITS
344 WAITS
345 WAITS
346 WAITS
347 WAITS
348 WAITS
349 WAITS
350 WAITS
351 WAITS
352 WAITS
353 WAITS
354 WAITS
355 WAITS
356 WAITS
357 WAITS
358 WAITS
359 WAITS
360 WAITS
361 WAITS
362 WAITS
363 WAITS
364 WAITS
365 WAITS
366 WAITS
367 WAITS
368 WAITS
369 WAITS
370 WAITS
371 WAITS
372 WAITS
373 WAITS
374 WAITS
375 WAITS
376 WAITS
377 WAITS
378 WAITS
379 WAITS
380 WAITS
381 WAITS
382 WAITS
383 WAITS
384 WAITS
385 WAITS
386 WAITS
387 WAITS
388 WAITS
389 WAITS
390 WAITS
391 WAITS
392 WAITS
393 WAITS
394 WAITS
395 WAITS
396 WAITS
397 WAITS
398 WAITS
399 WAITS
400 WAITS
401 WAITS
402 WAITS
403 WAITS
404 WAITS
405 WAITS
406 WAITS
407 WAITS
408 WAITS
409 WAITS
410 WAITS
411 WAITS
412 WAITS
413 WAITS
414 WAITS
415 WAITS
416 WAITS
417 WAITS
418 WAITS
419 WAITS
420 WAITS
421 WAITS
422 WAITS
423 WAITS
424 WAITS
425 WAITS
426 WAITS
427 WAITS
428 WAITS
429 WAITS
430 WAITS
431 WAITS
432 WAITS
433 WAITS
434 WAITS
435 WAITS
436 WAITS
437 WAITS
438 WAITS
439 WAITS
440 WAITS
441 WAITS
442 WAITS
443 WAITS
444 WAITS
445 WAITS
446 WAITS
447 WAITS
448 WAITS
449 WAITS
450 WAITS
451 WAITS
452 WAITS
453 WAITS
454 WAITS
455 WAITS
456 WAITS
457 WAITS
458 WAITS
459 WAITS
460 WAITS
461 WAITS
462 WAITS
463 WAITS
464 WAITS
465 WAITS
466 WAITS
467 WAITS
468 WAITS
469 WAITS
470 WAITS
471 WAITS
472 WAITS
473 WAITS
474 WAITS
475 WAITS
476 WAITS
477 WAITS
478 WAITS
479 WAITS
480 WAITS
481 WAITS
482 WAITS
483 WAITS
484 WAITS
485 WAITS
486 WAITS
487 WAITS
488 WAITS
489 WAITS
490 WAITS
491 WAITS
492 WAITS
493 WAITS
494 WAITS
495 WAITS
496 WAITS
497 WAITS
498 WAITS
499 WAITS
500 WAITS
501 WAITS
502 WAITS
503 WAITS
504 WAITS
505 WAITS
506 WAITS
507 WAITS
508 WAITS
509 WAITS
510 WAITS
511 WAITS
512 WAITS
513 WAITS
514 WAITS
515 WAITS
516 WAITS
517 WAITS
518 WAITS
519 WAITS
520 WAITS
521 WAITS
522 WAITS
523 WAITS
524 WAITS
525 WAITS
526 WAITS
527 WAITS
528 WAITS
529 WAITS
530 WAITS
531 WAITS
532 WAITS
533 WAITS
534 WAITS
535 WAITS
536 WAITS
537 WAITS
538 WAITS
539 WAITS
540 WAITS
541 WAITS
542 WAITS
543 WAITS
544 WAITS
545 WAITS
546 WAITS
547 WAITS
548 WAITS
549 WAITS
550 WAITS
551 WAITS
552 WAITS
553 WAITS
554 WAITS
555 WAITS
556 WAITS
557 WAITS
558 WAITS
559 WAITS
560 WAITS
561 WAITS
562 WAITS
563 WAITS
564 WAITS
565 WAITS
566 WAITS
567 WAITS
568 WAITS
569 WAITS
570 WAITS
571 WAITS
572 WAITS
573 WAITS
574 WAITS
575 WAITS
576 WAITS
577 WAITS
578 WAITS
579 WAITS
580 WAITS
581 WAITS
582 WAITS
583 WAITS
584 WAITS
585 WAITS
586 WAITS
587 WAITS
588 WAITS
589 WAITS
590 WAITS
591 WAITS
592 WAITS
593 WAITS
594 WAITS
595 WAITS
596 WAITS
597 WAITS
598 WAITS
599 WAITS
600 WAITS
601 WAITS
602 WAITS
603 WAITS
604 WAITS
605 WAITS
606 WAITS
607 WAITS
608 WAITS
609 WAITS
610 WAITS
611 WAITS
612 WAITS
613 WAITS
614 WAITS
615 WAITS
616 WAITS
617 WAITS
618 WAITS
619 WAITS
620 WAITS
621 WAITS
622 WAITS
623 WAITS
624 WAITS
625 WAITS
626 WAITS
627 WAITS
628 WAITS
629 WAITS
630 WAITS
631 WAITS
632 WAITS
633 WAITS
634 WAITS
635 WAITS
636 WAITS
637 WAITS
638 WAITS
639 WAITS
640 WAITS
641 WAITS
642 WAITS
643 WAITS
644 WAITS
645 WAITS
646 WAITS
647 WAITS
648 WAITS
649 WAITS
650 WAITS
651 WAITS
652 WAITS
653 WAITS
654 WAITS
655 WAITS
656 WAITS
657 WAITS
658 WAITS
659 WAITS
660 WAITS
661 WAITS
662 WAITS
663 WAITS
664 WAITS
665 WAITS
666 WAITS
667 WAITS
668 WAITS
669 WAITS
670 WAITS
671 WAITS
672 WAITS
673 WAITS
674 WAITS
675 WAITS
676 WAITS
677 WAITS
678 WAITS
679 WAITS
680 WAITS
681 WAITS
682 WAITS
683 WAITS
684 WAITS
685 WAITS
686 WAITS
687 WAITS
688 WAITS
689 WAITS
690 WAITS
691 WAITS
692 WAITS
693 WAITS
694 WAITS
695 WAITS
696 WAITS
697 WAITS
698 WAITS
699 WAITS
700 WAITS
701 WAITS
702 WAITS
703 WAITS
704 WAITS
705 WAITS
706 WAITS
707 WAITS
708 WAITS
709 WAITS
710 WAITS
711 WAITS
712 WAITS
713 WAITS
714 WAITS
715 WAITS
716 WAITS
717 WAITS
718 WAITS
719 WAITS
720 WAITS
721 WAITS
722 WAITS
723 WAITS
724 WAITS
725 WAITS
726 WAITS
727 WAITS
728 WAITS
729 WAITS
730 WAITS
731 WAITS
732 WAITS
733 WAITS
734 WAITS
735 WAITS
736 WAITS
737 WAITS
738 WAITS
739 WAITS
740 WAITS
741 WAITS
742 WAITS
743 WAITS
744 WAITS
745 WAITS
746 WAITS
747 WAITS
748 WAITS
749 WAITS
750 WAITS
751 WAITS
752 WAITS
753 WAITS
754 WAITS
755 WAITS
756 WAITS
757 WAITS
758 WAITS
759 WAITS
760 WAITS
761 WAITS
762 WAITS
763 WAITS
764 WAITS
765 WAITS
766 WAITS
767 WAITS
768 WAITS
769 WAITS
770 WAITS
771 WAITS
772 WAITS
773 WAITS
774 WAITS
775 WAITS
776 WAITS
777 WAITS
778 WAITS
779 WAITS
780 WAITS
781 WAITS
782 WAITS
783 WAITS
784 WAITS
785 WAITS
786 WAITS
787 WAITS
788 WAITS
789 WAITS
790 WAITS
791 WAITS
792 WAITS
793 WAITS
794 WAITS
795 WAITS
796 WAITS
797 WAITS
798 WAITS
799 WAITS
800 WAITS
801 WAITS
802 WAITS
803 WAITS
804 WAITS
805 WAITS
806 WAITS
807 WAITS
808 WAITS
809 WAITS
810 WAITS
811 WAITS
812 WAITS
813 WAITS
814 WAITS
815 WAITS
816 WAITS
817 WAITS
818 WAITS
819 WAITS
820 WAITS
821 WAITS
822 WAITS
823 WAITS
824 WAITS
825 WAITS
826 WAITS
827 WAITS
828 WAITS
829 WAITS
830 WAITS
831 WAITS
832 WAITS
833 WAITS
834 WAITS
835 WAITS
836 WAITS
837 WAITS
838 WAITS
839 WAITS
840 WAITS
841 WAITS
842 WAITS
843 WAITS
844 WAITS
845 WAITS
846 WAITS
847 WAITS
848 WAITS
849 WAITS
850 WAITS
851 WAITS
852 WAITS
853 WAITS
854 WAITS
855 WAITS
856 WAITS
857 WAITS
858 WAITS
859 WAITS
860 WAITS
861 WAITS
862 WAITS
863 WAITS
864 WAITS
865 WAITS
866 WAITS
867 WAITS
868 WAITS
869 WAITS
870 WAITS
871 WAITS
872 WAITS
873 WAITS
874 WAITS
875 WAITS
876 WAITS
877 WAITS
878 WAITS
879 WAITS
880 WAITS
881 WAITS
882 WAITS
883 WAITS
884 WAITS
885 WAITS
886 WAITS
887 WAITS
888 WAITS
889 WAITS
890 WAITS
891 WAITS
892 WAITS
893 WAITS
894 WAITS
895 WAITS
896 WAITS
897 WAITS
898 WAITS
899 WAITS
900 WAITS
901 WAITS
902 WAITS
903 WAITS
904 WAITS
905 WAITS
906 WAITS
907 WAITS
908 WAITS
909 WAITS
910 WAITS
911 WAITS
912 WAITS
913 WAITS
914 WAITS
915 WAITS
916 WAITS
917 WAITS
918 WAITS
919 WAITS
920 WAITS
921 WAITS
922 WAITS
923 WAITS
924 WAITS
925 WAITS
926 WAITS
927 WAITS
928 WAITS
929 WAITS
930 WAITS
931 WAITS
932 WAITS
933 WAITS
934 WAITS
935 WAITS
936 WAITS
937 WAITS
938 WAITS
939 WAITS
940 WAITS
941 WAITS
942 WAITS
943 WAITS
944 WAITS
945 WAITS
946 WAITS
947 WAITS
948 WAITS
949 WAITS
950 WAITS
951 WAITS
952 WAITS
953 WAITS
954 WAITS
955 WAITS
956 WAITS
957 WAITS
958 WAITS
959 WAITS
960 WAITS
961 WAITS
962 WAITS
963 WAITS
964 WAITS
965 WAITS
966 WAITS
967 WAITS
968 WAITS
969 WAITS
970 WAITS
971 WAITS
972 WAITS
973 WAITS
974 WAITS
975 WAITS
976 WAITS
977 WAITS
978 WAITS
979 WAITS
980 WAITS
981 WAITS
982 WAITS
983 WAITS
984 WAITS
985 WAITS
986 WAITS
987 WAITS
988 WAITS
989 WAITS
990 WAITS
991 WAITS
992 WAITS
993 WAITS
994 WAITS
995 WAITS
996 WAITS
997 WAITS
998 WAITS
999 WAITS
